airflow/utils/log/file_task_handler.py

```diff
@@ -0,0 +1,187 @@
+
+import logging
+import os
+
+from airflow import configuration as conf
+from airflow.configuration import AirflowConfigException
+from airflow.utils.file import mkdirs
+
+class FileTaskHandler(logging.Handler):
+    """
+    FileTaskHandler is a python log handler that handles and reads
+    task instance logs. It creates and delegates log handling
+    to `logging.FileHandler` after receiving task instance context.
+    It reads logs from task instance's host machine.
+    """
+
+    def __init__(self, base_log_folder):
+        super(FileTaskHandler, self).__init__()
+        self.handler = None
+        self.local_base = os.path.expanduser(base_log_folder)
+
+    def set_context(self, task_instance):
+        """
+        Provide task_instance context to airflow task handler.
+        :param task_instance: task instance object
+        """
+        local_loc = self._init_file(task_instance)
+        self.handler = logging.FileHandler(local_loc)
+        self.handler.setFormatter(self.formatter)
+        self.handler.setLevel(self.level)
+
+    def emit(self, record):
+        if self.handler:
+            self.handler.emit(record)
+
+    def flush(self):
+        if self.handler:
+            self.handler.flush()
+
+    def close(self):
+        if self.handler:
+            self.handler.close()
+
+    def _read(self, task_instance, try_number):
+        """
+        Template method that contains custom logic of reading
+        logs given the try_number.
+        :param task_instance: task instance record
+        :param try_number: current try_number to read log from
+        :return: log message as a string
+        """
+
+        # Task instance here might be different from task instance when
+        # initializing the handler. Thus explicitly getting log location
+        # is needed to get correct log path.
+        log_relative_path = self._get_log_relative_path(task_instance, try_number)
+        loc = os.path.join(self.local_base, log_relative_path)
+        log = ""
+
+        if os.path.exists(loc):
+            try:
+                with open(loc) as f:
+                    log += "*** Reading local log.\n" + "".join(f.readlines())
+            except Exception:
+                log = "*** Failed to load local log file: {0}.\n".format(loc)
+        else:
+            url = os.path.join("http://{ti.hostname}:{worker_log_server_port}/log",
+                               log_relative_path).format(
+                ti=task_instance,
+                worker_log_server_port=conf.get('celery', 'WORKER_LOG_SERVER_PORT'))
+            log += "*** Log file isn't local.\n"
+            log += "*** Fetching here: {url}\n".format(**locals())
+            try:
+                import requests
+                timeout = None  # No timeout
+                try:
+                    timeout = conf.getint('webserver', 'log_fetch_timeout_sec')
+                except (AirflowConfigException, ValueError):
+                    pass
```

Comment:
This is part of the refactor from the original code. Some of the logic might need to be changed, such as `mkdir`. I feel it's better to change in a separate PR, but I will definitely make a note here.